### ISSUE TITLE
Grain

### ISSUE THREAD

**Issue by Beep6581** [Beep6581/RawTherapee#2460]
Originally reported on Google Code with ID 2477

```
It would be useful to be able to add grain to the image, in some cases.
To make the noise-cleaned image appear more natural, with the controlled type of grain,
reduce posterization, for aesthetically pleasing effect, where applicable.

In Photoshop, I follow this simple technique via recorded action.
The result is more of "organic" looking grain.

Copy current layer
Work on the copy:
 Add Large grain pattern:
  - add gaussian noise monochromatic, strength 400%
  - gaussian blur, radius 1.1 pixels
 Add Small grain pattern:
  - add gaussian noise monochromatic, strength 4%
  - gaussian blur, radius 0.6 pixels
 Set Opacity (e.g. 17%)
Merge with original layer

It may also be useful to modulate opacity by luminance.

Grain should be added as one of the final treatments.
```

Reported by `michaelezra000` on 2014-08-24 15:51:20

**Comment by ror6ax** [comment]
+1. Often adding noise is the only thing I need to do outside of RT in editors like Gimp. Would've been great to have it.

**Comment by gigiboeru** [comment]
Noise would be a great addition! It's really useful in various circumstances...

**Comment by Lowkus** [comment]
Has there been any progress on this feature?  I really want to see it implemented.

**Comment by abnormally-distributed** [comment]
I agree, it would be great to have this. It's one of many features that darktable has that I wish rawtherapee had

**Comment by heckflosse** [comment]
You can add grain/noise using local adjustments in dev builds.
![grafik](https://user-images.githubusercontent.com/13733487/108206175-57c29d00-7126-11eb-899c-730363f8be82.png)

**Comment by Lowkus** [comment]
Where can we find pre-made dev builds?  Or do we need to compile it on our machines?

Is the solution applying noise using the correct logic of having less noise in the shadows and highlights.  There was a discussion about it here... https://discuss.pixls.us/t/lets-improve-grain/2709

**Comment by heckflosse** [comment]
> Where can we find pre-made dev builds?

https://github.com/Beep6581/RawTherapee/releases/tag/nightly

**Comment by Lowkus** [comment]
It's great that work is being made toward simulating film grain, and I'm excited about the potential for the spot tool.  In case they need some feedback about the grain simulation...
1) Although it is nice to have grain simulation available in the spot tool, for the sake of efficiency there also needs to be a way to apply grain to the entire image without using the spot tool.  Ideally it would be placed in the Film Simulation section near the LUT dropdown.
2) If the Courseness slider is meant to replicate film grain, there should be an option to lock the slider to traditional ISO increments such as 200, 400, 800.
3) The grain is not following the bell curve intensity mentioned in the previous article.  Highlights and shadows are being filled with grain that is too strong.
4) I'm not sure of the intention of the Courseness slider, it does not seem to impact the grain in any meaningful way.  It seemed to change the randomization seed of where grain was placed, but the grain size remained unaffected and the slider did not change how far apart the grains were placed.  In a more detailed grain simulation, it might be worth changing the program to show us these sliders...
  A) "Seed" (Randomization seed).
  B) "Strength" (Already implemented).
  C) "Spacing" (Distance between grains, statistically weighted.  I'm not sure if this would be worth the effort though, it sounds simple but the calculations could be difficult/slow).
  D) "Curve" (Amount of "Bell Curve" that is used when applying grain.  When set to 0 it would mean the grain has no bell curve logic, so the grain strength will be the same across the image.  When set to 100, the full bell curve is applied, making it so that highlights and shadows will have grain that is barely visible.  If desired the software could also take this setting to 200, which would do a sort of hyper tight bell curve where only the most middling of tones will get strong grain.
  E) "Center" (This would shift the central hump of the bell curve so it can be moved left or right.  Normally the curve is set to being in the middle, assigning the strongest grains to mid-tones like in real film.  But if we move the bell curve hump to the left or right we could make it so the strongest grain appears in the shadows, or in the highlights.  That could allow for some artistic choices beyond what would be accurate to real film.

**Comment by heckflosse** [comment]
@Desmis

**Comment by abnormally-distributed** [comment]
Nice. Do you happen to know when 5.9 will have a final release? Not sure if
the collaborator tag means you're one of the devs. I've had issues getting
a c++ compiler to work on my laptop (windows 10) for something else, so
I'll probably have to hold off on the dev build until public release

**Comment by Desmis** [comment]
@Lowkus   @heckflosse 

I ported the Darktable code (via ART), and I've just added a "hidden" slider - which I had hidden to simplify...
It is out of the question for me to retouch this code (either small adjustments)  :)
But if you or someone else wishes to modify it, no problem. In particular to make certain declared constants (const) "variable". 
https://github.com/Beep6581/RawTherapee/pull/6118

The fact that "grain" is in "Local Adjustments" (LA) makes it possible to use all the functionalities of LA, in particular :
* work with a spot or full image (selection  in « settings »)
* use deltaE (scope) to differentiate the action according to « colors  and luminance » (sky, skin, flowers, etc.)
* use mask 
* use excluding spot, to cancel actions - for some parts of the image
* use transitions
* and use the joint functionalities due to the design of "LA" - concept of object and possibility of multiple sessions

I don't know how to do with GUI, for a slider with incremental ISO...

jacques

**Comment by Desmis** [comment]
@Lowkus  @heckflosse 

As I mentioned, I tried with the Darktable algorithm to modify the parameters declared as "const".
After various tests, the only one that has a real influence is  GRAIN_LUT_PAPER_GAMMA and to a lesser degree GRAIN_LUT_DELTA_MAX ...other parameters are redundant

I modified the code of "ipgrain.cc" and the GUI code to take these changes into account.

Of course everything I said about "LA" is still true. 

I just open a PR
https://github.com/Beep6581/RawTherapee/pull/6120

I will be little available from this evening and for a week.

jacques
